specs/ChainSpec/Consensus.md

```diff
@@ -114,21 +118,34 @@ def get_approvals(self, target_height):
 
 A block producer assigned for a particular height produces a block at that height whenever they have `get_approvals` return approvals from block producers whose stake collectively exceeds 2/3 of the total stake.
 
-## Safety
+## Epoch Switches
+There's a parameter `epoch_length` in genesis config that defines the expected length of an epoch. Say a particular epoch `e_cur` started at height `h`, and say the next epoch will be `e_next`. Say `BP(e)` is a set of block producers in epoch `e`. Say `LFH(B)` is the height of most recent final block in the ancestry of `B`. The following are the rules of what blocks contain approvals from what block producers, and belong to what epoch.
 
-We call approvals for which `prev_height + 1 = target_height` endorsements, and all other approvals skips.
+- Any block `B` with height between `h` and `h + epoch_length - 3` (both inclusive) is in the epoch `e_cur` and must have approvals from more than 2/3 of `BP(e_cur)` (stake-weighted).
+- Any block `B` with height `h + epoch_length - 2` or higher for which `LFH(prev(B)) < h + epoch_length - 3` is in the epoch `e_cur` and must have the approvals from both more than 2/3 of `BP(e_cur)` and more than 2/3 of `BP(e_next)` (both stake-weighted).
+- The first block `B` with height `h + epoch_length - 2` or higher and `LFH(prev(B)) >= h + epoch_length - 3` is in the epoch `e_next` and only needs the approvals from more than 2/3 of `BP(e_next)` (stake-weighted).
```

Comment:
I assume that the intent is that under ideal conditions, the first block in `e_next` has height `h + epoch_length`, and it can never be less. Am I right? This implies that `LFH(B)` is defined such that for every `B`, `LFH(B).height <= B.height - 2`, which is absolutely not clear from the definition of `LFH`.